lib/permanent_not_found.rb

```diff
@@ -4,23 +4,28 @@
 class PermanentNotFound
   CONTENT = '404 Not Found'.freeze
 
-  def initialize(app, content: CONTENT, paths: [])
-    @app = app
+  def initialize(app, content: CONTENT, paths: [], pattern: [])
+    @app     = app
     @content = content
-    @paths = Set.new(paths)
+    @paths   = Set.new(paths)
+    @pattern = Regexp.union(pattern)
   end
 
   def call(env)
-    if @paths.include?(env['PATH_INFO'].downcase)
-      response
+    if permanent_redirected?(env['PATH_INFO'].downcase)
+      response_with_404
     else
       @app.call(env)
     end
   end
 
   private
 
-  def response
+  def permanent_redirected?(path)
+    @paths.include?(path) || @pattern =~ path
```

Comment:
Should this maybe be split into 2 further methods in order to clarify, that 2 different methods for verifying the inclusion of the path in the filter are used?